### ISSUE TITLE
Delegate creation fails to consider NotNullIfNotNull

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#44735]
**Version Used**: 16.7 Preview 1

**Steps to Reproduce**:

```csharp
#nullable enable

using System;
using System.Diagnostics.CodeAnalysis;

public class C {
    public Func<string?, string?> M1() {
        return Helper.Method;
    }

    public Func<string, string> M2() {
        return Helper.Method; // CS8620
    }
}

static class Helper {
    [return: NotNullIfNotNull("arg")]
    public static string? Method(string? arg) => arg;
}
```

**Expected Behavior**:

No warning is reported in `M2()`.

**Actual Behavior**:

CS8620 is reported in `M2()`.

**Comment by RikkiGibson** [comment]
related to #44129

**Comment by RikkiGibson** [comment]
We don't get this warning in an analogous override scenario either, where the "overridden" method is the delegate and the "overriding" method is the target method. (This works out because just like we call an overridden method and dispatch to its override, we call a delegate and dispatch to its target method.)

```cs
class Base1 {
    public virtual string? Method(string? arg) => arg;
}

class Derived1 : Base1 {
    [return: NotNullIfNotNull("arg")]
    public override string? Method(string? arg) => arg;
}

class Base2 {
    public virtual string Method(string arg) => arg;
}

class Derived2 : Base2 {
    [return: NotNullIfNotNull("arg")]
    public override string? Method(string? arg) => arg; // warning: nullability of return type doesn't match overridden member
}
```

[SharpLab link](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFBUACADITQIwB0AIgJYQDmaA9rA4BjKCwDCfACZwAgqSwBPKBygBuKrQBMhAELQ4TQgG8qhM4wDMhAG4cEMDBCyMmdAPyEAsnBgALKQAUzO6EEAjcAJSEALwAfKHh6pQAvhqUNNpsiBzWcJKGILr6hiaU5oQA2jQA7IUAcnwwddhYAJIAZg1NLQEARGHcvREAuqbmNFZ8uQgIHNIuId5+gcEeA1FxCdxJqdTp2npQcNql5RM2dg5OC14+/pJBrlsb8QM7aRmEWbO5ktqFh2OxjGZiqtUIXWaOA6kJ6/XCQ1GZXGk2ms3mq1uywemPWMVeiUIAHoiYQAO5hNAcNDcQqYHB4DhYDgwBSEPjtRjVQisgAOREkfDgUDQAHIYIRIDAhL52Wi5tI0JK4GBgIgqLsgA)

This is actually something new that I don't think we've tried to solve. I think we could solve this in override variance checks by making a pass over parameters with `NotNullIfNotNullAttribute`, then checking if the type of the base parameter should cause the overridden method's return to be considered "not null" for the purposes of variance checks, and then checking the compatibility between the base and derived "adjusted" return types. This is similar to what we do in `NullableWalker.VisitArguments`.

**Comment by jcouv** [comment]
@RikkiGibson Was this fixed by your recent change for nullability in delegate conversion (https://github.com/dotnet/roslyn/pull/46953) ?

**Comment by RikkiGibson** [comment]
This was not fixed. Details and proposed fix are outlined in https://github.com/dotnet/roslyn/issues/44735#issuecomment-675768560

**Comment by RikkiGibson** [comment]
I think this is a duplicate of #44174.
